beacon_node/beacon_chain/src/sync_committee_rewards.rs

```diff
@@ -0,0 +1,79 @@
+use crate::{BeaconChain, BeaconChainError, BeaconChainTypes};
+
+use eth2::lighthouse::SyncCommitteeReward;
+use state_processing::per_block_processing::altair::sync_committee::compute_sync_aggregate_rewards;
+use types::{BeaconBlockRef, BeaconState, ExecPayload};
+use std::collections::HashMap;
+
+impl<T: BeaconChainTypes> BeaconChain<T> {
+    pub fn compute_sync_committee_rewards<Payload: ExecPayload<T::EthSpec>>(
+        &self,
+        block: BeaconBlockRef<'_, T::EthSpec, Payload>,
+        state: &mut BeaconState<T::EthSpec>,
+    ) -> Result<Vec<SyncCommitteeReward>, BeaconChainError> {
+        if block.slot() != state.slot() {
+            return Err(BeaconChainError::BlockRewardSlotError);
+        }
+
+        let spec = &self.spec;
+
+        let sync_aggregate = block
+            .body()
+            .sync_aggregate()?;
+
+        let sync_committee = state
+            .current_sync_committee()?
+            .clone();
+
+        let sync_committee_indices = state
+            .get_sync_committee_indices(&sync_committee)?;
+
+        let (participant_reward_value, proposer_reward_per_bit) = compute_sync_aggregate_rewards(&state, spec)
+            .map_err(|_| BeaconChainError::SyncCommitteeRewardsSyncError)?;
+
+        let mut balances = sync_committee_indices
+            .iter()
+            .map(|i| (*i, state.balances()[*i]))
+            .collect::<HashMap<usize, u64>>();
+
+        let mut total_proposer_rewards = 0;
+        let proposer_index = state.get_beacon_proposer_index(block.slot(), spec)?;
+        balances.insert(proposer_index, state.balances()[proposer_index]);
+
+        // Apply rewards to participant balances. Keep track of proposer rewards
+        for (validator_index, participant_bit) in sync_committee_indices.iter().zip(sync_aggregate.sync_committee_bits.iter()) {
+            let participant_balance = balances.get(validator_index);
+
+            if participant_bit {
+                if let Some(balance_value) = participant_balance {
+                    balances.insert(*validator_index, balance_value + participant_reward_value);
+                }
+                *balances.get_mut(&proposer_index).unwrap() += proposer_reward_per_bit;
+                total_proposer_rewards +=  proposer_reward_per_bit;
+            } else {
+                if let Some(balance_value) = participant_balance {
+                    balances.insert(*validator_index, balance_value.saturating_sub(participant_reward_value));
+                }
+            }
+        }
```

Comment:
Usually we try to avoid `unwrap` because panics will crash the thread (and usually the whole node, outside the HTTP API).

One way to work around this is to use the `HashMap::entry` API, and do the initialization of balances inside the loop:

```rust
        // Apply rewards to participant balances. Keep track of proposer rewards
        for (&validator_index, participant_bit) in sync_committee_indices
            .iter()
            .zip(sync_aggregate.sync_committee_bits.iter())
        {
            let participant_balance = balances
                .entry(validator_index)
                .or_insert_with(|| state.balances()[validator_index]);

            if participant_bit {
                participant_balance.safe_add_assign(participant_reward_value)?;

                balances
                    .entry(proposer_index)
                    .or_insert_with(|| state.balances()[proposer_index])
                    .safe_add_assign(proposer_reward_per_bit)?;

                total_proposer_rewards.safe_add_assign(proposer_reward_per_bit)?;
            } else {
                *participant_balance = participant_balance.saturating_sub(participant_reward_value);
            }
        }
```

This is how I re-worked it just now, also using `safe_arith::SafeArith` to get the `safe_add_assign` method which is kind of useful (and we guard against overflow, which is nice but probably not super necessary outside of consensus).

Reply:
I understand balances may not have every validator_index that you are looping hence we need to `.or_insert_with(|| state.balances()[validator_index]);`
However for proposer_index, since I already explicitly inserted proposer’s balance into balance before the loop `balances.insert(proposer_index, state.balances()[proposer_index]);` it is mostly guaranteed that `balances.entry(proposer_index)` will return non-empty entry. Is it necessary to have `or_insert_with` on `proposer_index` ?

Reply:
Yeah, generally we just avoid `unwrap()` completely. Even if an error _shouldn't_ be reachable, that might get messed up in a future refactor, and an error result is _always_ better than a panic. Plus by outlawing `unwrap` it makes it easier to use lints like this one: https://rust-lang.github.io/rust-clippy/master/index.html#unwrap_used (we use that lint in `consensus/`).

For this could you could remove the `insert`s entirely and allow the map to be lazily populated by the `or_insert_with`, i.e. create the map completely empty and don't do _any_ inserts except from `or_insert_with`. Another option would be to add a new error like `RewardsLogicError` and do:

```rust
balances.get_mut(&proposer_index).ok_or(Error::RewardsLogicError)?;
```

This is a pattern we use in a bunch of other places -- unlikely errors just get a basic error variant which we expect never to see.